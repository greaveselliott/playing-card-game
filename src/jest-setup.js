// Make Enzyme functions available in all test files without importing
import { shallow, render, mount } from 'enzyme';
import React from 'react';

global.shallow = shallow;
global.render = render;
global.mount = mount;

// Fail tests on any warning
console.error = message => {
    throw new Error(message);
};

/**
 * Function to mock an Enzyme component.
 * @param enzymeRender {function} render function.
 * @param Component {JSXComponent} component type to mock.
 * @param componentProps {Object} optional component properties.
 * @param reduxStoreContext {Object} optional redux store context.
 * @returns Enzyme Component.
 */
global.mockComponent = function (enzymeRender, Component, componentProps = {}, reduxStoreContext = {}) {

    let jsxElement = <Component {...componentProps} />;
    const enzymeComponent = enzymeRender(jsxElement, reduxStoreContext);

    return {
        componentProps,
        enzymeComponent
    }
};

/**
 * Function to create data driven tests against poly component configurations
 * @param Component {JSXComponent} component type to mock.
 * @param configurations {Array of objects} An object containing the test configuration for each test case. Each object within this array must have a 'description' and 'props' property.
 * @returns nothing
 */
global.assertComponentProperties = function (Component, configurations) {

    const expectedProperties = ['description', 'props'];

    configurations.forEach( (testCase) => {

        expectedProperties.forEach((property) => {
            if(!testCase.hasOwnProperty(property)) {
                throw new Error(`Error: assertComponentProperties 'configuration' property does not have required '${property}' property.`);
            }
        });

        const { enzymeComponent } = mockComponent(shallow, Component, testCase.props);

        it(testCase.description, () => {
            expect(enzymeComponent).toMatchSnapshot();
        });
    });
};

global.mockResponse = function (status, statusText, response) {
    return new window.Response(response, {
        status: status,
        statusText: statusText,
        headers: {
            'Content-type': 'application/json'
        }
    });
};