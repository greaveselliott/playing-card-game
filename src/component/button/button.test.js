import React from 'react';
import Button from './button';
import '../../jest-setup';

describe("Button component", () => {  

    it("Should render a Button component.", () => {
        const props = {
            text : "Label",
        };

        const { enzymeComponent } = mockComponent(shallow, Button, props);   

        expect(enzymeComponent).toMatchSnapshot();
    });

    it("Should render a Button component with a modifier.", () => {
        const props = {
            text : "Label",
            modifier: "--active"
        };
        
        const { enzymeComponent } = mockComponent(shallow, Button, props);   

        expect(enzymeComponent).toMatchSnapshot();
    });

    it("Should render a Button component with a additional class.", () => {
        const props = {
            text : "Label",
            modifier: "--active",
            class_name: "an addition class"
        };
        
        const { enzymeComponent } = mockComponent(shallow, Button, props);   

        expect(enzymeComponent).toMatchSnapshot();
    });

    it("Should render a Button component with a click event handler.", () => {
        const spy = jest.fn();
        const props = {
            text : "Label",
            modifier: "--active",
            class_name: "an addition class",
            click_handler: spy
        };
        const { enzymeComponent } = mockComponent(shallow, Button, props);   

        enzymeComponent.simulate('click');
        expect(spy).toHaveBeenCalled();
    });
});