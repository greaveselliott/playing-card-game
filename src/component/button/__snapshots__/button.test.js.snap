// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Should render a Button component with a additional class. 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <button
    className="a-button--active an addition class"
    onClick={[Function]}
>
    Label
</button>,
  "nodes": Array [
    <button
      className="a-button--active an addition class"
      onClick={[Function]}
>
      Label
</button>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Button
        class_name="an addition class"
        modifier="--active"
        text="Label"
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "class_name": "an addition class",
          "modifier": "--active",
          "text": "Label",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <button
          className="a-button--active an addition class"
          onClick={[Function]}
>
          Label
</button>,
        "_debugID": 6,
        "_renderedOutput": <button
          className="a-button--active an addition class"
          onClick={[Function]}
>
          Label
</button>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Button
    class_name="an addition class"
    modifier="--active"
    text="Label"
/>,
}
`;

exports[`Button component Should render a Button component with a modifier. 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <button
    className="a-button--active "
    onClick={[Function]}
>
    Label
</button>,
  "nodes": Array [
    <button
      className="a-button--active "
      onClick={[Function]}
>
      Label
</button>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Button
        modifier="--active"
        text="Label"
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "modifier": "--active",
          "text": "Label",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <button
          className="a-button--active "
          onClick={[Function]}
>
          Label
</button>,
        "_debugID": 4,
        "_renderedOutput": <button
          className="a-button--active "
          onClick={[Function]}
>
          Label
</button>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Button
    modifier="--active"
    text="Label"
/>,
}
`;

exports[`Button component Should render a Button component. 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <button
    className="a-button "
    onClick={[Function]}
>
    Label
</button>,
  "nodes": Array [
    <button
      className="a-button "
      onClick={[Function]}
>
      Label
</button>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Button
        text="Label"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "text": "Label",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <button
          className="a-button "
          onClick={[Function]}
>
          Label
</button>,
        "_debugID": 2,
        "_renderedOutput": <button
          className="a-button "
          onClick={[Function]}
>
          Label
</button>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Button
    text="Label"
/>,
}
`;
